File: ObjC/AoC2025/AoC2025/Utility/AOCGrid.m
//
//  AOCGrid2D.m
//  AoC2025
//
//  Created by Simon Biickert on 2023-01-29.
//

#import <Foundation/Foundation.h>
#import "AOCGrid.h"
#import "AOCStrings.h"

@implementation AOCGrid {
	NSMutableDictionary<AOCCoord *, NSObject *> *_data;
	AOCExtent *_extent;
}

+ (AOCGrid *)grid
{
	return [[AOCGrid alloc] init];
}

- (AOCGrid *)init
{
	self = [self initWithDefault:@"." adjacency:ROOK];
	return self;
}
- (AOCGrid *)initWithDefault:(NSObject *)value adjacency:(NSString *)rule
{
	self = [super init];
	
	_data = [NSMutableDictionary dictionary];
	_extent = nil;
	_defaultValue = value;
	_rule = rule;
	
	return self;
}

- (void)load:(NSArray<NSString *> *)data {
	[self clearAll];
	for (NSInteger row = 0; row < data.count; row++) {
		NSArray<NSString *> *letters = data[row].allCharacters;
		for (NSInteger col = 0; col < letters.count; col++) {
			[self setObject:letters[col] atCoord:[AOCCoord x:col y:row]];
		}
	}
}

- (NSObject *)objectAtCoord:(AOCCoord *)coord
{
	AOCCoord *c = coord;
	if (self.isTiledInfinitely && self.extent != nil && ![_extent contains:coord]) {
		NSInteger x = coord.x; NSInteger y = coord.y;
		while (x < _extent.min.x) { x += _extent.width; }
		while (y < _extent.min.y) { y += _extent.height; }
		while (x > _extent.max.x) { x -= _extent.width; }
		while (y > _extent.max.y) { y -= _extent.height; }
		c = [AOCCoord x:x y:y];
	}
	if (c == nil || [_data objectForKey:c] == nil) {
		return self.defaultValue;
	}
	return [_data objectForKey:c];
}

- (NSInteger)integerAtCoord:(AOCCoord *)coord
{
	NSNumber *num = (NSNumber *)[self objectAtCoord:coord];
	return [num integerValue];
}

- (NSString *)stringAtCoord:(AOCCoord *)coord
{
	NSString *str;
	id obj = [self objectAtCoord:coord];
	if ([obj isKindOfClass:[NSString class]]) {
		str = (NSString *)[self objectAtCoord:coord];
	}
	else if ([obj respondsToSelector:@selector(glyph)]) {
		str = [obj glyph];
	}
	else {
		str = [NSString stringWithFormat:@"%@", obj];
	}
	return str;
}


- (void)setObject:(NSObject *)value atCoord:(AOCCoord *)coord
{
	[_data setObject:value forKey:coord];
	if (_extent == nil) {
		_extent = [[AOCExtent alloc] initFrom:@[coord]];
	}
	else {
		[_extent expandToFit:coord];
	}
}

- (void)clearAtCoord:(AOCCoord *)coord {
	[_data removeObjectForKey:coord];
}

- (void)clearAtCoord:(AOCCoord *)coord resetExtent:(BOOL)reset {
	[self clearAtCoord: coord];
	if (reset) {
		_extent = [[AOCExtent alloc] initFrom:self.coords];
	}
}

- (void)clearAll
{
	[_data removeAllObjects];
	_extent = nil;
}

- (AOCExtent *)extent
{
	if (!_extent) { return nil; }
	return [AOCExtent copyOf:_extent];
}

- (NSArray<AOCCoord *> *)coords
{
	return [_data allKeys];
}

- (NSArray<AOCCoord *> *)coordsWithValue:(NSObject *)value
{
	NSMutableArray<AOCCoord *> *coords = [NSMutableArray array];
	for (AOCCoord *key in [_data allKeys]) {
		if ([[_data objectForKey:key] isEqualTo:value]) {
			[coords addObject:key];
		}
	}
	return coords;
}

- (NSDictionary<NSString *, NSNumber *> *)histogram {
	NSMutableDictionary<NSString *, NSNumber *> *result = [NSMutableDictionary dictionary];
	
	for (AOCCoord *c in self.coords) {
		NSString *value = [self stringAtCoord:c];
		if (![result objectForKey:value]) {
			result[value] = @0;
		}
		result[value] = [NSNumber numberWithInteger:[result[value] integerValue] + 1];
	}
	
	return result;
}

- (NSDictionary<NSString *, NSNumber *> *)histogramIncludingDefaults {
	NSMutableDictionary<NSString *, NSNumber *> *result = [NSMutableDictionary dictionary];
	
	AOCExtent *ext = self.extent;
	if (ext) {
		for (NSInteger row = ext.min.y; row <= ext.max.y; row++) {
			for (NSInteger col = ext.min.x; col <= ext.max.x; col++) {
				NSString *value = [self stringAtCoord:[AOCCoord x:col y:row]];
				if (![result objectForKey:value]) {
					result[value] = @0;
				}
				result[value] = [NSNumber numberWithInteger:[result[value] integerValue] + 1];
			}
		}
	}
	
	return result;
}


- (NSArray<AOCCoord *> *)adjacentTo:(AOCCoord *)coord;
{
	NSMutableArray<AOCCoord *> *a = [NSMutableArray array];
	
	NSArray<AOCCoord *> *o = [AOCCoord adjacentOffsetsWithRule:self.rule];
	for (AOCCoord *offset in o) {
		[a addObject:[coord add:offset]];
	}
	
	return a;
}

- (NSArray<AOCCoord *> *)adjacentTo:(AOCCoord *)coord withValue:(NSObject *)value
{
	NSArray<AOCCoord *> *allAdjacent = [self adjacentTo:coord];
	NSMutableArray<AOCCoord *> *result = [NSMutableArray array];
	
	for (AOCCoord *coord in allAdjacent) {
		if ([[_data objectForKey:coord] isEqualTo:value]) {
			[result addObject:coord];
		}
	}
	
	return result;
}

- (void)print
{
	[self printWithOverlay:nil drawExtent:nil];
}

- (void)printWithOverlay:(NSDictionary<AOCCoord *, NSString *> *)overlay {
	[self printWithOverlay:overlay drawExtent:nil];
}

- (void)printWithOverlay:(NSDictionary<AOCCoord *, NSString *> *)overlay 
			  drawExtent:(AOCExtent *)drawExtent {
	NSString *str = [self toStringWithOverlay:overlay
								   drawExtent:drawExtent];
	[str println];
}

- (NSString *)toString {
	return [self toStringWithOverlay:nil drawExtent:nil];
}

- (NSString *)toStringWithOverlay:(NSDictionary<AOCCoord *, NSString *> *)overlay {
	return [self toStringWithOverlay:overlay drawExtent:nil];
}

- (NSString *)toStringWithOverlay:(NSDictionary<AOCCoord *, NSString *> *)overlay 
					   drawExtent:(AOCExtent *)drawExtent {
	
	NSMutableString *str = [NSMutableString string];
	AOCExtent *ext = self.extent;
	
	if (self.isTiledInfinitely && drawExtent != nil) {
		ext = drawExtent;
	}
		
	for (NSInteger row = ext.min.y; row <= ext.max.y; row++) {
		NSMutableString *line = [NSMutableString string];
		for (NSInteger col = ext.min.x; col <= ext.max.x; col++) {
			AOCCoord *c = [AOCCoord x:col y:row];
			NSString *s = [self stringAtCoord:c];
			
			if (overlay != nil && [overlay objectForKey:c] != nil) {
				s = [overlay objectForKey:c];
			}
			
			[line appendString:s];
		}
		[str appendFormat:@"%@\n", line];
	}

	return str;
}



- (NSDictionary<AOCCoord *, NSObject *> *)data
{
	return [_data copy];
}

- (BOOL)isEqualToGrid:(AOCGrid *)other
{
	return (self.defaultValue == other.defaultValue &&
			[self.rule isEqualToString:other.rule] &&
			[_data isEqualTo:other.data]);
}


@end

File: Swift/Advent of Code 2024.swiftpm/Libraries/AoCGrid.swift
//
//  AoCGrid2D.swift
//  AoC 2023
//
//  Created by Simon Biickert on 2023-11-27.
//

import Foundation

class AoCGrid2D {
    let defaultValue: String
    private var _data = Dictionary<AoCCoord2D, Any>()
    private var _adjacencyRule: AoCAdjacencyRule = .rook
    private var _extent: AoCExtent2D?
    var isTiledInfinitely: Bool = false
    
    init(defaultValue: String = ".", rule: AoCAdjacencyRule = .rook) {
        self.defaultValue = defaultValue
        _adjacencyRule = rule
        _extent = nil
    }
    
    func load(data: [String]) {
        for row in 0..<data.count {
            for col in 0..<data[row].count {
                let s = String(data[row][col])
                if s != defaultValue {
                    self.setValue(s, at: AoCCoord2D(x: col, y: row))
                }
            }
        }
    }
    
    var rule: AoCAdjacencyRule {
        return _adjacencyRule
    }
    
    var extent: AoCExtent2D? {
        return _extent
    }
    
    func stringValue(at coord: AoCCoord2D) -> String {
        let val = value(at: coord)
        return AoCGrid2D.transformToString(value: val)
    }
    
    private static func transformToString(value: Any) -> String {
        if let str = value as? String {
            return str
        }
        else if let renderable = value as? AoCGridRenderable {
            return renderable.glyph
        }
        return "\(value)"
        
    }
    
    func value(at coord: AoCCoord2D) -> Any {
        var c = coord
        if isTiledInfinitely && extent != nil && extent!.contains(coord) == false {
            // This coord is outside the bounds of the grid. What would a tiled copy of the grid return?
            c = AoCCoord2D(x: AoCUtil.trueMod(num: coord.x, mod: extent!.width),
                           y: AoCUtil.trueMod(num: coord.y, mod: extent!.height))
        }
        if let v = _data[c] {
            return v
        }
        return defaultValue
    }
    
    func setValue(_ v: Any, at coord: AoCCoord2D) {
        _data[coord] = v
        if extent == nil {
            _extent = AoCExtent2D(min: coord, max: coord)
        }
        else {
            _extent = extent!.expanded(toFit: coord)
        }
    }
    
    func fill(with value:String, at point:AoCCoord2D, filled: inout [AoCCoord2D]) -> Bool {
        if !extent!.contains(point) {
            return true
        }
        var touchedInfinity = false
        setValue(value, at: point)
        filled.append(point)
        let neighbours = neighbourCoords(at: point, withValue: defaultValue)
        for neighbour in neighbours {
            touchedInfinity = touchedInfinity || fill(with: value, at: neighbour, filled: &filled)
        }
        if touchedInfinity {
            for filledPoint in filled {
                clear(at: filledPoint)
            }
        }
        return touchedInfinity
    }
    
    func clear(at coord: AoCCoord2D, resetExtent: Bool = false) {
        _data.removeValue(forKey: coord)
        if (resetExtent) {
            _extent = AoCExtent2D.build(from: coords)
        }
    }
    
    var coords: [AoCCoord2D] {
        return Array(_data.keys)
    }
    
    var values: [Any] {
        return Array(_data.values)
    }
    
    func getCoords(withValue v: String) -> [AoCCoord2D] {
        let result = _data.filter {
            let str = AoCGrid2D.transformToString(value: $0.value)
            return str == v
        }
        return Array(result.keys)
    }
    
    var histogram: Dictionary<String, Int> {
        var result = Dictionary<String, Int>()
        if let ext = extent {
            for row in ext.min.y...ext.max.y {
                for col in ext.min.x...ext.max.x {
                    let v = stringValue(at: AoCCoord2D(x: col, y: row))
                    if result.keys.contains(v) == false { result[v] = 0 }
                    result[v]! += 1
                }
            }
        }
        return result
    }
    
    var neighbourOffsets: [AoCCoord2D] {
        return AoCCoord2D.getAdjacentOffsets(rule: self.rule)
    }
    
    func neighbourCoords(at coord: AoCCoord2D) -> [AoCCoord2D] {
        return coord.getAdjacentCoords(rule: self.rule)
    }
    
    func neighbourCoords(at coord: AoCCoord2D, withValue s: String) -> [AoCCoord2D] {
        var result = neighbourCoords(at: coord)
        result = result.filter { self.stringValue(at: $0) == s }
        return result
    }
    
    func toString(markers: Dictionary<AoCCoord2D, String>? = nil, drawExtent: AoCExtent2D? = nil) -> String {
        var str = ""
        if var ext = extent {
            
            if isTiledInfinitely && drawExtent != nil { ext = drawExtent! }
            
            for row in ext.min.y...ext.max.y {
                var values = [String]()
                for col in ext.min.x...ext.max.x {
                    let coord = AoCCoord2D(x: col, y: row)
                    if let markers = markers,
                       markers.keys.contains(coord) {
                        values.append(markers[coord]!)
                    }
                    else {
                        values.append(stringValue(at: coord))
                    }
                }
                str.append(values.joined(separator: " "))
                str.append("\n")
            }
        }
        return str
    }
    
    func draw(markers: Dictionary<AoCCoord2D, String>? = nil, drawExtent: AoCExtent2D? = nil) {
        print(self.toString(markers: markers, drawExtent: drawExtent))
    }
}

protocol AoCGridRenderable {
    var glyph: String {get}
}

